src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts

```
@@ -0,0 +1,517 @@
+               * For outside passed permission modes, it may contains other permissions. Add a intersection
+               * here to make sure only Library related permission modes will be used.
+               */
+              permissionModes: options.ACLSearchParams.permissionModes
+                ? intersection(options.ACLSearchParams.permissionModes, [
+                    WorkspacePermissionMode.LibraryRead,
+                    WorkspacePermissionMode.LibraryWrite,
+                  ])
+                : [WorkspacePermissionMode.LibraryRead, WorkspacePermissionMode.LibraryWrite],
+            },
+          })
+        ).saved_objects.map((item) => item.id);
+
+        if (options.workspaces) {
+          const permittedWorkspaces = options.workspaces.filter((item) =>
+            (permittedWorkspaceIds || []).includes(item)
+          );
+          if (!permittedWorkspaces.length) {
+            /**
+             * If user does not have any one workspace access
+             * deny the request
+             */
+            throw SavedObjectsErrorHelpers.decorateNotAuthorizedError(
+              new Error(
+                i18n.translate('workspace.permission.invalidate', {
+                  defaultMessage: 'Invalid workspace permission',
+                })
+              )
+            );
+          }
+
+          /**
+           * Overwrite the options.workspaces when user has access on partial workspaces.
+           */
+          options.workspaces = permittedWorkspaces;
+        } else {
+          /**
+           * Select all the docs that
+           * 1. ACL matches read / write / user passed permission OR
+           * 2. workspaces matches library_read or library_write OR
+           * 3. Advanced settings
+           */
+          options.workspaces = undefined;
+          options.ACLSearchParams.workspaces = permittedWorkspaceIds;
+          options.ACLSearchParams.permissionModes = options.ACLSearchParams.permissionModes ?? [
```

Could `options.ACLSearchParams.permissionModes` be an empty array `[]`? What's the consequence when `permissionModes = []` is passed in options?

---

Discussed with @wanglam offline, just for transparency, when listing workspaces, passing empty array makes it to return all workspaces including those workspaces that the current user doesn't have access to.

---

Sure, add fail fast check in generate search dsl function. https://github.com/wanglam/OpenSearch-Dashboards/pull/4/files#diff-428b7d9ca0d879eea511a66eec222df247a1fe9ff32e13feb7ab53677b6ff044